### PR TITLE
version-service - Move command deduplication feature descriptors into the experimental features [kvl-1218]

### PR DESCRIPTION
**Pull request by nicu-da**
- As the command deduplication descriptors are used for testing purposes only, they should be set in the message of the experimental feature

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 1]
```diff
@@ -17,7 +17,38 @@ option csharp_namespace = "Com.Daml.Ledger.Api.V1";
...
+  }
+  OffsetSupport offset_support = 1;
+  DurationSupport duration_support = 2;
+}
+
+// Ledger support for participant deduplication
+enum ParticipantDeduplicationSupport {
+  PARTICIPANT_DEDUPLICATION_NOT_SUPPORTED = 0;
+  PARTICIPANT_DEDUPLICATION_PARALLEL_ONLY = 1;
+  PARTICIPANT_DEDUPLICATION_SUPPORTED = 2;
```
Comment (fabiotudone-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:53]:
```suggestion
  // The participant does not support deduplication.
  PARTICIPANT_DEDUPLICATION_NOT_SUPPORTED = 0;
  // The participant only supports deduplicating submissions that are being concurrently processed.
  PARTICIPANT_DEDUPLICATION_PARALLEL_ONLY = 1;
  // The participant supports deduplication for all the submissions it receives.
  PARTICIPANT_DEDUPLICATION_SUPPORTED = 2;
```

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 2]
```diff
@@ -17,7 +17,38 @@ option csharp_namespace = "Com.Daml.Ledger.Api.V1";
...
+    // Ledgers use offsets for command deduplication
+    OFFSET_NATIVE_SUPPORT = 0;
+    // Ledger convert offsets into a duration for command deduplication
+    OFFSET_CONVERT_TO_DURATION = 1;
+    // Ledger does not support offsets for command deduplication
+    OFFSET_NOT_SUPPORTED = 2;
+  }
+  enum DurationSupport {
+    DURATION_NATIVE_SUPPORT = 0;
+    DURATION_CONVERT_TO_OFFSET = 1;
```
Comment (fabiotudone-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:43]:
```suggestion
    // The ledger natively support deduplication periods specified as durations.
    DURATION_NATIVE_SUPPORT = 0;
    // The ledger doesn't natively support deduplication periods specified as durations but it converts them into offsets.
    DURATION_CONVERT_TO_OFFSET = 1;
```

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 3]
```diff
@@ -17,7 +17,38 @@ option csharp_namespace = "Com.Daml.Ledger.Api.V1";
...
+
+
+message CommandDeduplicationFeatures {
+  DeduplicationPeriodSupport deduplication_period_support = 1;
+  ParticipantDeduplicationSupport participant_deduplication_support = 2;
+}
+
+message DeduplicationPeriodSupport {
+  enum OffsetSupport {
+    // Ledgers use offsets for command deduplication
```
Comment (fabiotudone-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:34]:
```suggestion
    // The ledger natively supports deduplication periods specified as offsets.
```

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 4]
```diff
@@ -17,7 +17,38 @@ option csharp_namespace = "Com.Daml.Ledger.Api.V1";
...
+  CommandDeduplicationFeatures command_deduplication = 2;
 }
 
 // GRPC self-service error codes are returned by the Ledger API.
 message ExperimentalSelfServiceErrorCodes {}
+
+
+message CommandDeduplicationFeatures {
+  DeduplicationPeriodSupport deduplication_period_support = 1;
+  ParticipantDeduplicationSupport participant_deduplication_support = 2;
```
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:29]: How do these fields interact? Probably not all values are sensible. Could you add comments that clarify this?

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 5]
```diff
@@ -17,7 +17,38 @@ option csharp_namespace = "Com.Daml.Ledger.Api.V1";
...
+    OFFSET_CONVERT_TO_DURATION = 1;
+    // Ledger does not support offsets for command deduplication
+    OFFSET_NOT_SUPPORTED = 2;
+  }
+  enum DurationSupport {
+    DURATION_NATIVE_SUPPORT = 0;
+    DURATION_CONVERT_TO_OFFSET = 1;
+  }
+  OffsetSupport offset_support = 1;
+  DurationSupport duration_support = 2;
```
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:46]: How do these fields interact? Probably not all values are sensible. Could you add comments that clarify this?

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 6]
```diff
@@ -17,7 +17,38 @@ option csharp_namespace = "Com.Daml.Ledger.Api.V1";
...
+message CommandDeduplicationFeatures {
+  DeduplicationPeriodSupport deduplication_period_support = 1;
+  ParticipantDeduplicationSupport participant_deduplication_support = 2;
+}
+
+message DeduplicationPeriodSupport {
+  enum OffsetSupport {
+    // Ledgers use offsets for command deduplication
+    OFFSET_NATIVE_SUPPORT = 0;
+    // Ledger convert offsets into a duration for command deduplication
```
Comment (fabiotudone-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:36]:
```suggestion
    // The ledger doesn't natively support deduplication periods specified as offsets but it converts them into durations.
```

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 7]
```diff
@@ -17,7 +17,38 @@ option csharp_namespace = "Com.Daml.Ledger.Api.V1";
...
+    OFFSET_CONVERT_TO_DURATION = 1;
+    // Ledger does not support offsets for command deduplication
```
Comment (fabiotudone-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:38]:
```suggestion
    // The ledger does not support support specifiying deduplication periods as offsets.
```

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto (hunk 1 above)
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:53]: What do these values mean? Add comments :)

**Review by meiersi-da — CHANGES_REQUESTED** [review]

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 8]
```diff
@@ -17,7 +17,38 @@ option csharp_namespace = "Com.Daml.Ledger.Api.V1";
...
 message ExperimentalFeatures {
   ExperimentalSelfServiceErrorCodes self_service_error_codes = 1;
+  CommandDeduplicationFeatures command_deduplication = 2;
 }
 
 // GRPC self-service error codes are returned by the Ledger API.
 message ExperimentalSelfServiceErrorCodes {}
+
+
+message CommandDeduplicationFeatures {
```
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:27]: Consider adding a comment that provides context as to the reason for these features, their stability, and their target state.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 9]
```diff
@@ -17,7 +17,38 @@ option csharp_namespace = "Com.Daml.Ledger.Api.V1";
 // See the feature message definitions for descriptions.
 message ExperimentalFeatures {
   ExperimentalSelfServiceErrorCodes self_service_error_codes = 1;
+  CommandDeduplicationFeatures command_deduplication = 2;
 }
 
 // GRPC self-service error codes are returned by the Ledger API.
 message ExperimentalSelfServiceErrorCodes {}
+
```
Comment (fabiotudone-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:25]: Nit:

```suggestion
```

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto (hunk 1 above)
Comment (nicu-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto]: Thank you @fabiotudone-da for adding the descriptions

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto (hunk 5 above)
Comment (nicu-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto]: The fields themselves don't really interact. There's nothing preventing a ledger from having any type of offset support in combination with any type of duration support.

**Review by hubert-da — APPROVED** [review]
LGTM

**Comment by nicu-da** [comment]
@simonmaxen-da added a few more changes based on our chat, [`0767d6f` (#12318)](https://github.com/digital-asset/daml/pull/12318/commits/0767d6f4532221bc496546641f231c6478bf2aba)

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 10]
```diff
@@ -53,11 +68,11 @@ message DeduplicationPeriodSupport {
 }
 
 // Ledger support for participant deduplication
-enum ParticipantDeduplicationSupport {
+enum DeduplicationType {
   // The participant does not support deduplication.
```
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:72]: @nicu-da : Out of date comment? here and below?

type is helpful though :)

**Review by meiersi-da — APPROVED** [review]

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 11]
```diff
@@ -17,7 +17,43 @@ option csharp_namespace = "Com.Daml.Ledger.Api.V1";
...
+    DURATION_NATIVE_SUPPORT = 0;
+    DURATION_CONVERT_TO_OFFSET = 1;
+  }
+  OffsetSupport offset_support = 1;
+  DurationSupport duration_support = 2;
+}
+
+// How the participant node reports duplicate command submissions.
+enum CommandDeduplicationType {
+  // Duplicate commands are exclusively reported asynchronous via completions.
```
Comment (fabiotudone-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:52]:
```suggestion
  // Duplicate commands are exclusively reported asynchronously via completions.
```

**Review by fabiotudone-da — APPROVED** [review]

**Review by SamirTalwar-DA — APPROVED** [review]

File: ledger/sandbox-classic/src/main/scala/platform/sandbox/SandboxServer.scala [hunk 12]
```diff
@@ -428,12 +428,13 @@ final class SandboxServer(
...
-              offsetSupport = DeduplicationPeriodSupport.OffsetSupport.OFFSET_NOT_SUPPORTED,
-              durationSupport = DeduplicationPeriodSupport.DurationSupport.DURATION_NATIVE_SUPPORT,
+            CommandDeduplicationPeriodSupport.of(
+              offsetSupport = CommandDeduplicationPeriodSupport.OffsetSupport.OFFSET_NOT_SUPPORTED,
+              durationSupport =
+                CommandDeduplicationPeriodSupport.DurationSupport.DURATION_NATIVE_SUPPORT,
             )
           ),
-          ParticipantDeduplicationSupport.PARTICIPANT_DEDUPLICATION_SUPPORTED,
+          CommandDeduplicationType.ASYNC_ONLY,
```
Comment (SamirTalwar-DA) [inline ledger/sandbox-classic/src/main/scala/platform/sandbox/SandboxServer.scala:437]:
```suggestion
          CommandDeduplicationType.SYNC_ONLY,
```

Comment (nicu-da) [reply ledger/sandbox-classic/src/main/scala/platform/sandbox/SandboxServer.scala]: Yup, at least we know the feature descriptors work as expected based on all the failed tests.

**Pull request merged**